api/utils/pkixname/parser.go

```diff
@@ -0,0 +1,583 @@
+	}
+
+	tokens, err := tokenize(dn)
+	if err != nil {
+		return nil, err
+	}
+
+	dst := &pkix.Name{}
+	if tokens.Len() == 0 {
+		return dst, nil
+	}
+	if err := parseRDNSequence(dst, *tokens); err != nil {
+		return nil, fmt.Errorf("malformed RDNs: %w", err)
+	}
+	return dst, nil
+}
+
+// parseRDNSequence parses a RelativeDistinguishedName sequence, ie, a sequence
+// of AttributeTypeAndValue separated by commas or pluses.
+func parseRDNSequence(dst *pkix.Name, tokens tokenList) error {
+	seenAttrs := make(map[string]struct{})
+	markAttr := func(attr string) error {
+		if _, ok := seenAttrs[attr]; ok {
+			return fmt.Errorf("repeated attributeType %q, remaining tokens: %s", attr, tokens)
+		}
+		seenAttrs[attr] = struct{}{}
+		return nil
+	}
+
+	prevAttr, err := parseATV(dst, tokens)
+	if err != nil {
+		return err
+	}
+	_ = markAttr(prevAttr)
+
+	for {
+		tok, ok := tokens.Peek()
+		if !ok {
+			return nil // end
+		}
+
+		switch tok.kind {
+		case tokenPlus:
+			tokens.PopSilently()
+
+			// Validate that prevAttr == current attr.
+			// If `!ok` just keep going and parseATV() will fail.
+			if nextTok, ok := tokens.Peek(); ok &&
+				nextTok.kind == tokenAttrType &&
+				nextTok.value != prevAttr {
+				return fmt.Errorf(
+					"multi-valued RDN must refer to the same attribute, but found %q instead of %q, remaining tokens: %s",
+					prevAttr,
+					nextTok.value,
+					tokens,
+				)
+			}
+
+			prevAttr, err = parseATV(dst, tokens)
+			if err != nil {
+				return err
+			}
+
+		case tokenComma:
+			tokens.PopSilently()
+			prevAttr, err = parseATV(dst, tokens)
+			if err != nil {
+				return err
+			}
+			if err := markAttr(prevAttr); err != nil {
+				return err
+			}
+
+		default:
+			// Force an error.
+			return requireTokenKind(tokenComma, tok, tokens)
+		}
+	}
+}
+
+// parseATV parses an AttributeTypeAndValue.
+//
+// Eg: "CN" EQUAL "Llama CA".
+func parseATV(dst *pkix.Name, tokens tokenList) (attr string, _ error) {
+	t1, t2, t3, ok := tokens.Peek3()
+	if !ok {
+		return "", fmt.Errorf(
+			"not enough tokens to parse AttributeTypeValue, remaining tokens: %s",
+			tokens,
+		)
+	}
+	if err := requireTokenKind(tokenAttrType, t1, tokens); err != nil {
+		return "", err
+	}
+	if err := requireTokenKind(tokenEqual, t2, tokens); err != nil {
+		return "", err
+	}
+	if err := requireTokenKind(tokenString, t3, tokens); err != nil {
+		return "", err
+	}
+
+	// Pop tokens before returning. We retain the tokens up until the end so
+	// eventual errors include them in the message.
+	defer func() {
+		tokens.PopSilently()
+		tokens.PopSilently()
+		tokens.PopSilently()
+	}()
+
+	attr = t1.value
+	value := t3.value
+
+	// Parse as OID?
+	if oidRegexp.MatchString(attr) {
+		return attr, parseOIDExtraName(dst, attr, value)
+	}
+	// Verify attributeType character set.
+	if !attrTypeRegexp.MatchString(attr) {
+		return "", fmt.Errorf("invalid attributeType (bad character set): %q", attr)
+	}
+
+	switch attr {
```

Comment:
This seems strictly case sensitive, should we ignore case on attribute types?

Reply:
I think being case sensitive here is OK. (It might even be correct? I'm not sure.)